src/vmm/src/lib.rs

```diff
@@ -324,32 +310,30 @@ impl VMM {
         mptable::setup_mptable(&self.guest_memory, vcpu_cfg.num_vcpus)
             .map_err(|e| Error::Vcpu(vcpu::Error::Mptable(e)))?;
 
-        let base_cpuid = self
-            .kvm
-            .get_supported_cpuid(KVM_MAX_CPUID_ENTRIES)
-            .map_err(Error::KvmIoctl)?;
-
         // Safe to use expect() because the device manager is instantiated in new(), there's no
         // default implementation, and the field is private inside the VMM struct.
         let shared_device_mgr = Arc::new(self.device_mgr.take().expect("Missing device manager"));
 
         for index in 0..vcpu_cfg.num_vcpus {
-            let vcpu =
-                Vcpu::new(&self.vm_fd, index, shared_device_mgr.clone()).map_err(Error::Vcpu)?;
-
             // Set CPUID.
+            self.vm.create_vcpu(index, shared_device_mgr.clone())?;
+        }
+        // Set CPUID.
+        let base_cpuid = self
+            .kvm
+            .get_supported_cpuid(KVM_MAX_CPUID_ENTRIES)
+            .map_err(Error::KvmIoctl)?;
+        for vcpu in self.vm.vcpus.iter() {
```

Comment:
While it will greatly simplify things in this particular use case, we would still need to offer functionality for setting a custom cpuid because the filtering is application specific. If we keep things as they are now then Firecracker would be able to consume the vm-vcpu crate and provide a cpuid that corresponds to the Firecracker specific templates (T2/C3). I would rather prefer to have something like creating a default cpuid to simplify the code here instead of moving this call in `Vm`. What do you say?

Reply:
The input (i.e. cpuid config) we use here to create each vcpu could be passed to `KvmVm` as well so it can handle things behind the scenes. For now we can proceed with either approach and see how things shape up.